### ISSUE TITLE
DefaultRefinement

### ISSUE THREAD

**Issue by davidfurlong** [algolia/react-instantsearch#61]
**Feature Request**

Currently its not clear how `defaultRefinement` behaves - it should be clearer that reseting / removing the refinement with `refine(null)` (or equivalent) will not remove the refinement if the `defaultRefinement` is set, but rather reset the refinement to the `defaultRefinement`.

Additionally I have use for an `initialRefinement` which acts like defaultRefinement except will actually be removed when refine(null) is called. I know there are ways to do initialRefinement now, but its not immediate how one should best implement it (to minimize API calls)

Edit: Additionally its not clear when a refinement is reset or not if we use defaultRefinement (as checking that the currentRefinement === defaultRefinement is not sufficient to know that its reset) (also the case with connectRange falling back to max & min values in lieu of defaultRefinement, when its not provided)

**Comment by mthuret** [comment]
I'm interesting to see how you handle this use case with the library at the moment? Can you share a bit of code? I'm also interesting by the business use case behind it :)

Thanks!

**Comment by davidfurlong** [comment]
@mthuret https://github.com/DeedMob/react-instantsearch-datepicker-widget/blob/master/DatePicker.js

Used currently like this:

```js
<ConnectedDatePicker
   attributeName="date_starting"
   min={getToday()}
   max={getYearFromNow()}
 />
```

**Comment by mthuret** [comment]
That's true we do not remove `defaultRefinement` when clearing current refinements (whether from `<ClearAll>`, `<CurrentRefinements`> or by using the connectCurrentRefinements with `refine(null)`. That way we differentiate all the filters a user could add and wants to remove, from some default configuration that you may enforce. You're right we should be clearer in our doc.

That being said I'm not sure you need to use the `connectCurrentRefinements` to actually clear your calendar. Why not just use the refine function of the `connectRange`? 

```js
<button
            onClick={() =>
              (this.props.defaultRefinement
                ? this.props.refine({
                    min: this.props.defaultRefinement.min,
                    max: this.props.defaultRefinement.max,
                  })
                : this.props.refine({
                    min: this.props.min,
                    max: this.props.max,
                  }))}
>
            Reset
</button>
```

Here if I have `defaultRefinement`, I fallback to them, otherwise I use the values automatically calculated by Algolia.

**Comment by davidfurlong** [comment]
You're right about the `connectCurrentRefinements` of course, that code needs to be refactored. 

I posted this issue because I struggled to initially "remove a refinement" - because a) defaultRefinement has this unexpected behaviour and b) theres no call that removes a refinement - setting values to null doesnt seem like its the same as removing the refinement. With objects as refinement values, i.e. connectRange, `refine(null)` throws an error and `refine({min: null, max: null})` is what works, but its not obvious nor documented clearly
